### PR TITLE
SUBMARINE-500. Add metric param instance to the submarine-server rest

### PR DESCRIPTION
**Pull request by JohnTing**
### What is this PR for?
Add  metric, param instance from submarine/store/database/models.py  to the submarine-server rest api

A few sentences describing the overall goals of the pull request's commits.
First time? Check out the contributing guide - https://submarine.apache.org/contribution/contributions.html

### What type of PR is it?
Feature 

### Todos
* [x] - metric rest api
* [x] - param rest api
* [x] - database

### What is the Jira issue?
https://issues.apache.org/jira/projects/SUBMARINE/issues/SUBMARINE-500

### How should this be tested?
https://travis-ci.org/github/JohnTing/submarine/builds/698241108

### Screenshots (if appropriate)
image
image
image
image
image
image

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? Yes/No

### CODE REVIEW

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+      e.printStackTrace();
+      return new JsonResponse.Builder<String>(Response.Status.OK).success(true).result("error").build();
+    }
+    return new JsonResponse.Builder<List<Metric>>(Response.Status.OK).success(true).result(metrics).build();
+  }
+
+  @GET
+  @Path("/")
+  @SubmarineApi
+  public Response getMetric(@QueryParam("id") String id) {
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:105]: Maybe changing id to path variable would be better.
like., 
```
@Path("/{id}")
@SubmarineApi
public Response getMetric(@PahtParam("id") String id) {
```

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    Metric metric;
+    try {
+      metric = metricService.selectById(id);
+
+    } catch (Exception e) {
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:111]: delete this blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+      LOG.error(e.toString());
+      e.printStackTrace();
+      return new JsonResponse.Builder<String>(Response.Status.OK).success(true).result("error").build();
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:114]: `Response.Status.OK` means the rest request status was successful.
`success(true)` or `success(flse)` indicates success or failure of the operation.
So, Here should be changed to
```
return new JsonResponse.Builder<String>(Response.Status.OK).success(false).build();
```

Other places in this PR also need to be revised.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+  @SubmarineApi
+  public Response postMetric(Metric metric) {
+    boolean result = false;
+    try {
+      result = metricService.insert(metric);
+    } catch (Exception e) {
+
+      LOG.error(e.toString());
+      e.printStackTrace();
+      return new JsonResponse.Builder<String>(Response.Status.OK).success(true).result("error").build();
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:130]: Change to `.success(false).build()`.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+  @Path("/delete")
+  @SubmarineApi
+  public Response deleteMetric(@QueryParam("id") String id) {
+    boolean result = false;
+    try {
+      result = metricService.deleteById(id);
+    } catch (Exception e) {
+      LOG.error(e.toString());
+      e.printStackTrace();
+      return new JsonResponse.Builder<String>(Response.Status.OK).success(true).result("error").build();
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:145]: Change to `.success(false).build()`.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+  @POST
+  @Path("/selective")
+  @SubmarineApi
+  public Response selectByPrimaryKeySelective(Metric metric) {
+    List<Metric> metrics;
+    try {
+      metrics = metricService.selectByPrimaryKeySelective(metric);
+
+    } catch (Exception e) {
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:174]: delete this blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  @POST
+  @Path("/selective")
+  @SubmarineApi
+  public Response selectByPrimaryKeySelective(Metric metric) {
+    List<Metric> metrics;
+    try {
+      metrics = metricService.selectByPrimaryKeySelective(metric);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:172]: delete this blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    } catch (Exception e) {
+
+      LOG.error(e.toString());
+      e.printStackTrace();
+      return new JsonResponse.Builder<String>(Response.Status.OK).success(true).result("error").build();
+    }
+    return new JsonResponse.Builder<List<Metric>>(Response.Status.OK).success(true).result(metrics).build();
+  }
+
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:182]: delete this blank lline.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    LOG.info("listParam ({})", param);
+    
+    List<Param> params;
+    try {
+      params = paramService.selectByPrimaryKeySelective(param);
+
+    } catch (Exception e) {
+
+      LOG.error(e.toString());
+      e.printStackTrace();
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:87]: Don't use `e.printStackTrace()` print log.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+      return new JsonResponse.Builder<String>(Response.Status.OK).success(true).result("error").build();
+    }
+    return new JsonResponse.Builder<List<Param>>(Response.Status.OK).success(true).result(params).build();
+  }
+
+  @GET
+  @Path("/")
+  @SubmarineApi
+  public Response getParam(@QueryParam("id") String id) {
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:96]: Maybe changing `id` to path variable would be better.
like., 
```
@Path("/{id}")
@SubmarineApi
public Response getParam(@PahtParam("id") String id) {
```

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    LOG.info("getParam ({})", id);
+    
+    Param param;
+    try {
+      param = paramService.selectById(id);
+
+    } catch (Exception e) {
+
+      LOG.error(e.toString());
+      e.printStackTrace();
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:106]: Don't use e.printStackTrace() print log.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+  @SubmarineApi
+  public Response postParam(Param param) {
+    LOG.info("postParam ({})", param);
+    boolean result = false;
+    try {
+      result = paramService.insert(param);
+    } catch (Exception e) {
+
+      LOG.error(e.toString());
+      e.printStackTrace();
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:123]: Don't use e.printStackTrace() print log.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+  @Path("/edit")
+  @SubmarineApi
+  public Response putParam(Param param) {
+    LOG.info("putParam ({})", param);
+    boolean result = false;
+    try {
+      result = paramService.update(param);
+    } catch (Exception e) {
+      LOG.error(e.toString());
+      e.printStackTrace();
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:155]: Don't use e.printStackTrace() print log.

File: submarine-server/server-core/src/main/resources/org/apache/submarine/database/mappers/MetricMapper.xml [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+      timestamp = #{timestamp,jdbcType=BIGINT},
+      step = #{step,jdbcType=INTEGER},
+      is_nan = #{isNan,jdbcType=INTEGER},
+      job_name = #{jobName,jdbcType=VARCHAR}
+    where id = #{id,jdbcType=VARCHAR}
+  </update>
+  
+
+
+    <select id="selectByPrimaryKeySelective" parameterType="java.lang.String" resultMap="resultMap">
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/resources/org/apache/submarine/database/mappers/MetricMapper.xml:86]: 2 spaces need to be indented

File: submarine-server/server-core/src/main/resources/org/apache/submarine/database/mappers/MetricMapper.xml [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+      worker_index = #{workerIndex,jdbcType=VARCHAR},
+      timestamp = #{timestamp,jdbcType=BIGINT},
+      step = #{step,jdbcType=INTEGER},
+      is_nan = #{isNan,jdbcType=INTEGER},
+      job_name = #{jobName,jdbcType=VARCHAR}
+    where id = #{id,jdbcType=VARCHAR}
+  </update>
+  
+
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/resources/org/apache/submarine/database/mappers/MetricMapper.xml:85]: Keep only one blank line

File: submarine-server/server-core/src/test/java/org/apache/submarine/server/workbench/database/service/MetricServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+    Metric metric = new Metric();
+    metric.setMetricKey("score");
+    metric.setValue((float) 0.666667);
+    metric.setWorkerIndex("worker-1");
+    metric.setTimestamp(new BigInteger("1569139525097"));
+    metric.setStep(0);
+    metric.setIsNan(0);
+    metric.setJobName("application_1234");
+    boolean result = metricService.insert(metric);
+    assertNotEquals(result, -1);
```
Comment (liuxunorg) [inline submarine-server/server-core/src/test/java/org/apache/submarine/server/workbench/database/service/MetricServiceTest.java:58]: result is boolean type, It should be compared to `true` or `false`, not to `-1`

assertTrue(result);

File: submarine-server/server-core/src/test/java/org/apache/submarine/server/workbench/database/service/ParamServiceTest.java [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+
+  @Test
+  public void testSelect() throws Exception {
+    Param param = new Param();
+    param.setParamKey("score");
+    param.setValue("199");
+    param.setWorkerIndex("worker-1");
+    param.setJobName("application_123651651");
+    boolean result = paramService.insert(param);
+    assertNotEquals(result, -1);
```
Comment (liuxunorg) [inline submarine-server/server-core/src/test/java/org/apache/submarine/server/workbench/database/service/ParamServiceTest.java:54]: change to `assertTrue(result);`

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/service/MetricService.java [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+    return true;
+  }
+
+  public Metric selectById(String id) throws Exception {
+    LOG.info("Metric selectByPrimaryKey {}", id);
+    Metric metric;
+    try (SqlSession sqlSession = MyBatisUtil.getSqlSession()) {
+      MetricMapper mapper = sqlSession.getMapper(MetricMapper.class);
+      metric = mapper.selectById(id);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/service/MetricService.java:89]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/service/MetricService.java [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  public List<Metric> selectByPrimaryKeySelective(Metric metric) throws Exception {
+    List<Metric> result;
+    LOG.info("Metric selectByPrimaryKeySelective");
+
+    try (SqlSession sqlSession = MyBatisUtil.getSqlSession()) {
+      MetricMapper mapper = sqlSession.getMapper(MetricMapper.class);
+      result = mapper.selectByPrimaryKeySelective(metric);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/service/MetricService.java:118]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/service/ParamService.java [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+  
+  private static final Logger LOG = LoggerFactory.getLogger(ParamService.class);
+
+  public List<Param> selectAll() throws Exception {
+    LOG.info("Param selectAll");
+    List<Param> params;
+    try (SqlSession sqlSession = MyBatisUtil.getSqlSession()) {
+      ParamMapper mapper = sqlSession.getMapper(ParamMapper.class);
+      params = mapper.selectAll();
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/service/ParamService.java:40]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/service/ParamService.java [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+  }
+  
+  public boolean deleteById(String id) throws Exception {
+    LOG.info("Param deleteById {}", id);
+
+    try (SqlSession sqlSession = MyBatisUtil.getSqlSession()) {
+      ParamMapper mapper = sqlSession.getMapper(ParamMapper.class);
+      mapper.deleteById(id);
+      sqlSession.commit();
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/service/ParamService.java:55]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/service/ParamService.java [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+    return true;
+  }
+  
+  public Param selectById(String id) throws Exception {
+    LOG.info("Param selectById {}", id);
+    Param param;
+    try (SqlSession sqlSession = MyBatisUtil.getSqlSession()) {
+      ParamMapper mapper = sqlSession.getMapper(ParamMapper.class);
+      param = mapper.selectById(id);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/service/ParamService.java:83]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/service/ParamService.java [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+  }
+
+  public List<Param> selectByPrimaryKeySelective(Param param) throws Exception {
+    List<Param> result;
+    LOG.info("Param selectByPrimaryKeySelective");
+
+    try (SqlSession sqlSession = MyBatisUtil.getSqlSession()) {
+      ParamMapper mapper = sqlSession.getMapper(ParamMapper.class);
+      result = mapper.selectByPrimaryKeySelective(param);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/service/ParamService.java:112]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 24]
```diff
@@ -0,0 +1,181 @@
...
+    LOG.info("listMetric ({})", metric);
+
+    List<Metric> metrics;
+    try {
+      metrics = metricService.selectByPrimaryKeySelective(metric);
+
+    } catch (Exception e) {
+      LOG.error(e.toString());
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:96]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 25]
```diff
@@ -0,0 +1,181 @@
...
+    metric.setIsNan(isNan);
+    metric.setJobName(jobName);
+    metric.setId(id);
+
+    LOG.info("listMetric ({})", metric);
+
+    List<Metric> metrics;
+    try {
+      metrics = metricService.selectByPrimaryKeySelective(metric);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:93]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 26]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  @GET
+  @Path("/{id}")
+  @SubmarineApi
+  public Response getMetric(@PathParam("id") String id) {
+    Metric metric;
+    try {
+      metric = metricService.selectById(id);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:109]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 27]
```diff
@@ -0,0 +1,181 @@
...
+    } catch (Exception e) {
+      LOG.error(e.toString());
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:112]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 28]
```diff
@@ -0,0 +1,181 @@
...
+    LOG.info("listMetric ({})", metric);
+
+    List<Metric> metrics;
+    try {
+      metrics = metricService.selectByPrimaryKeySelective(metric);
+
+    } catch (Exception e) {
+      LOG.error(e.toString());
+
+      return new JsonResponse.Builder<Boolean>(Response.Status.OK).success(true).result(false).build();
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:97]:
```
return new JsonResponse.Builder<Boolean>(Response.Status.OK).success(true).result(false).build();
```
1. Response.Status.OK: Indicates that the rest request was successful
2. success(true) or success(false): Indicates whether the background processing is normal or abnormal
3. result(false): Indicates that the background processing is normal, then the processing result is returned.
If the background is abnormal, there is no need to return any value. The caller determines whether there is a return value based on the success value.

SO，Please change 
```
return new JsonResponse.Builder<Boolean>(Response.Status.OK).success(true).result(false).build();
```
to
```
return new JsonResponse.Builder<Boolean>(Response.Status.OK).success(false).build();
```

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+  @Path("/add")
+  @SubmarineApi
+  public Response postMetric(Metric metric) {
+    boolean result = false;
+    try {
+      result = metricService.insert(metric);
+
+    } catch (Exception e) {
+      LOG.error(e.toString());
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:128]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 30]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  @POST
+  @Path("/add")
+  @SubmarineApi
+  public Response postMetric(Metric metric) {
+    boolean result = false;
+    try {
+      result = metricService.insert(metric);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:125]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 31]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  @DELETE
+  @Path("/delete")
+  @SubmarineApi
+  public Response deleteMetric(@QueryParam("id") String id) {
+    boolean result = false;
+    try {
+      result = metricService.deleteById(id);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:141]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 32]
```diff
@@ -0,0 +1,181 @@
...
+    } catch (Exception e) {
+      LOG.error(e.toString());
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:144]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 33]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  @PUT
+  @Path("/edit")
+  @SubmarineApi
+  public Response putMetric(Metric metric) {
+    boolean result = false;
+    try {
+      result = metricService.update(metric);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:157]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 34]
```diff
@@ -0,0 +1,181 @@
...
+    } catch (Exception e) {
+      LOG.error(e.toString());
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:160]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java [hunk 35]
```diff
@@ -0,0 +1,181 @@
...
+  @Path("/selective")
+  @SubmarineApi
+  public Response selectByPrimaryKeySelective(Metric metric) {
+    List<Metric> metrics;
+    try {
+      metrics = metricService.selectByPrimaryKeySelective(metric);
+
+    } catch (Exception e) {
+      LOG.error(e.toString());
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/MetricRestApi.java:176]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 36]
```diff
@@ -0,0 +1,177 @@
...
+    param.setValue(value);
+    param.setWorkerIndex(workerIndex);
+    param.setJobName(jobName);
+
+    LOG.info("listParam ({})", param);
+    
+    List<Param> params;
+    try {
+      params = paramService.selectByPrimaryKeySelective(param);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:84]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 37]
```diff
@@ -0,0 +1,177 @@
...
+    } catch (Exception e) {
+      LOG.error(e.toString());
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:87]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 38]
```diff
@@ -0,0 +1,177 @@
...
+  @GET
+  @Path("/{id}")
+  @SubmarineApi
+  public Response getParam(@PathParam("id") String id) {
+    LOG.info("getParam ({})", id);
+    
+    Param param;
+    try {
+      param = paramService.selectById(id);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:102]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 39]
```diff
@@ -0,0 +1,177 @@
...
+  @POST
+  @Path("/add")
+  @SubmarineApi
+  public Response postParam(Param param) {
+    LOG.info("postParam ({})", param);
+    boolean result = false;
+    try {
+      result = paramService.insert(param);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:119]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 40]
```diff
@@ -0,0 +1,177 @@
...
+    } catch (Exception e) {
+      LOG.error(e.toString());
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:122]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 41]
```diff
@@ -0,0 +1,177 @@
...
+  @DELETE
+  @Path("/delete")
+  @SubmarineApi
+  public Response deleteParam(@QueryParam("id") String id) {
+    LOG.info("deleteParam ({})", id);
+    boolean result = false;
+    try {
+      result = paramService.deleteById(id);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:136]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 42]
```diff
@@ -0,0 +1,177 @@
...
+    } catch (Exception e) {
+      LOG.error(e.toString());
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:139]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 43]
```diff
@@ -0,0 +1,177 @@
...
+  @PUT
+  @Path("/edit")
+  @SubmarineApi
+  public Response putParam(Param param) {
+    LOG.info("putParam ({})", param);
+    boolean result = false;
+    try {
+      result = paramService.update(param);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:153]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 44]
```diff
@@ -0,0 +1,177 @@
...
+    } catch (Exception e) {
+      LOG.error(e.toString());
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:156]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 45]
```diff
@@ -0,0 +1,177 @@
...
+  }
+
+  @POST
+  @Path("/selective")
+  @SubmarineApi
+  public Response selectByPrimaryKeySelective(Param metric) {
+    List<Param> params;
+    try {
+      params = paramService.selectByPrimaryKeySelective(metric);
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:169]: Please delete this a blank line.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java [hunk 46]
```diff
@@ -0,0 +1,177 @@
...
+    } catch (Exception e) {
+      LOG.error(e.toString());
+
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ParamRestApi.java:172]: Please delete this a blank line.

File: submarine-server/server-core/src/main/resources/org/apache/submarine/database/mappers/MetricMapper.xml [hunk 47]
```diff
@@ -0,0 +1,105 @@
...
+    </if>
+    <if test="workerIndex != null">
+      AND worker_index = #{workerIndex,jdbcType=INTEGER}
+    </if>
+    <if test="jobName != null">
+      AND job_name = #{jobName,jdbcType=VARCHAR}
+    </if>
+    <if test="id != null">
+      AND id = #{id,jdbcType=VARCHAR}
+    </if>
```
Comment (liuxunorg) [inline submarine-server/server-core/src/main/resources/org/apache/submarine/database/mappers/MetricMapper.xml:103]: Need Indent 2 spaces.

File: docs/database/submarine.sql [hunk 48]
```diff
@@ -219,3 +219,32 @@ CREATE TABLE `job` (
...
+  `is_nan` int(11) NOT NULL COMMENT 'True if the value is in fact NaN.',
+  `job_name` varchar(32) NOT NULL COMMENT 'JOB NAME to which this metric belongs to: Part of *Primary Key* for ``metrics`` table.',
+  PRIMARY KEY  (`id`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
+
+-- ----------------------------
+-- Table structure for param
+-- ----------------------------
+DROP TABLE IF EXISTS `param`;
+CREATE TABLE `param` (
```
Comment (pingsutw) [inline docs/database/submarine.sql:243]:
```suggestion
DROP TABLE IF EXISTS `params`;
CREATE TABLE `params` (
```
pysubmarine tracking API will log parameters to the table named `params`. so here we need to change to `params`

File: docs/database/submarine.sql [hunk 49]
```diff
@@ -219,3 +219,32 @@ CREATE TABLE `job` (
   `update_time` datetime default NULL COMMENT 'last update time',
   PRIMARY KEY  (`id`)
   ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
+  
+-- ----------------------------
+-- Table structure for metric
+-- ----------------------------
+DROP TABLE IF EXISTS `metric`;
+CREATE TABLE `metric` (
```
Comment (pingsutw) [inline docs/database/submarine.sql:227]:
```suggestion
DROP TABLE IF EXISTS `metrics`;
CREATE TABLE `metrics` (
```

**Comment by liuxunorg** [comment]
hi @JohnTing , Have you completed all the changes?

**Comment by JohnTing** [comment]
> hi @JohnTing , Have you completed all the changes?

Yes, all the changes are completed.

**Pull request closed without merging**